### PR TITLE
reverse bridge: support getting message length from upstream header

### LINKED ISSUE #12957: grpc_http1_reverse_bridge: support getting message length from upstream header
*Title*: *grpc_http1_reverse_bridge: support getting message length from upstream header*

*Description*:
Right now the gRPC reverse bridge filter can operate in 2 modes based on the withhold_grpc_frames config option. If it's false, then the upstream must prefix gRPC messages with the message-length frame header. If it's true, then the filter buffers the entire upstream response so it can count the length and add the prefix itself.

I think it would be useful to have a 3rd option which would be to get the message length from an upstream header, which would allow the upstream to avoid having knowledge of the gRPC message format, but still let the filter stream rather than buffer.

One weird wrinkle is we may not want to use the usual `content-length` header, because the upstream may want to use a `transfer-encoding: chunked` still which means that the `content-length` header should not be set.

Proposal:

Add a string `response_size_header` option, which lets users specify a custom header to use to get the length in bytes of the response message. If this is set, ignore `withhold_grpc_frames`.

### PR DESCRIPTION
**Pull request by baojr**
Commit Message: Add a configuration option to use the value of an upstream-supplied
response header as the content length to allow streaming responses.
Risk Level: Low
Testing: Integration test for happy path and unit tests
Docs Changes: Inline documentation on new proto field
Release Notes: One line mention in version history doc
Platform Specific Features: N/A

Fixes #12957

### CODE REVIEW

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:38]: This is confusing which option is being referred to; is the option `reponse_size_header`? This isn't a boolean so can't be true..

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:36]: Nit: s/envoy/Envoy/g

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:41]: I feel this whole explanation does not make sense to someone not deeply familiar with the context. Can you rewrite for an audience who does not have this?

**Review by htuch — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (baojr) [reply api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:41]: Thanks, I'll take another pass on the API comments tomorrow.

**Comment by alyssawilk** [comment]
Mark can you do an API pass while Harvey's out?

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (alyssawilk) [inline api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:47]: I think it's great to have an option to have this filter stream, but we should be careful to have Envoy validate the content length and fail if it's not correct, WDYT?

If so let's add code and tests for incorrect content length, and document the failure modes both for this, and if the response size header is missing ("the upstream must  set the header with this name -> & or Envoy will treat the response as an error")

File: docs/root/version_history/current.rst
Comment (alyssawilk) [inline docs/root/version_history/current.rst:66]: I think you can reflink to the config field here

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:157]: FWIW this does not handle multiple inline headers 

so you'd parse
Length: 4
Length 4
as 4 but not
Length: 4, 4
which is also valid.  Is that intentional?

File: test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc
Comment (alyssawilk) [inline test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc:247]: That's a really useful example.  Think it's worth calling out in API docs?

Though curious, if the endpoint knows the size (which it must in order to send the header) would it be possible to not have it chunk and just skip all this complexity altogether?  Largely content length is left out if the sender doesn't know the size of the response, which isn't true in this case.

**Review by alyssawilk — COMMENTED** [review]
Sorry for the lag - I'll pick up review while Lizan is out.

File: api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/grpc_http1_reverse_bridge/v3/config.proto:41]: Nit: `*response_size_header*` (and below)

**Review by htuch — COMMENTED** [review]
API LGTM modulo @alyssawilk comments and nit.

File: test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc
Comment (baojr) [reply test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc:247]: That's a good point. I'm going to confirm that the person asking for this feature doesn't need to chunk.

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (baojr) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:157]: I wrote in the API docs that only the first header would be used if multiple were present, since I can't think of a reasonable way to interpret repeated headers. Does that sound right to you? The other option I can think of would be to fail immediately, but that seems too strict to me.

It sounds like I need to read up on the spec a little bit, since I'm not familiar with repeated inline headers, but I plan to update the code to only use the first value in all cases.

**Comment by baojr** [comment]
Thanks for the API feedback! I will aim to push another update early next week.

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (snowp) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:160]: I don't think you need both of these?

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (snowp) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:168]: Maybe use Internal instead of Unknown?

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.h
Comment (snowp) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.h:30]: i think you can use `absl::make_optional` to avoid the type repetition here

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (snowp) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:216]: Just thinking out loud: perhaps it would be simpler to call addEncodedData in encodeHeaders to inject the frame header  and then simply pass through all the data in encodeData? I'm not sure if this is a better approach (just alternate), so unless there is some real benefit to redoing it like that we can keep things as is

File: test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc
Comment (snowp) [inline test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc:38]: I think we usually use `fmt::format`, which takes `{}` instead of `%s`

File: test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc
Comment (snowp) [inline test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_integration_test.cc:172]: Nice catch

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks pretty good overall, just a few comments

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by snowp** [comment]
Needs main merge and DCO fix

/wait

**Comment by baojr** [comment]
> looks like this still needs some format/docs tweaks :-)

I've been busy with my day job, but I plan to fix the failures today. I seem to have missed something in setting up my MacOS dev environment, so I'm working through that now.

**Comment by snowp** [comment]
Seems like this is still failing CI

/wait

**Comment by baojr** [comment]
Still working on the format check. The script is passing locally, so I haven't figured it out quite yet.

**Comment by baojr** [comment]
Thanks for everyone's patience while I figure out the build issues. I have a better understanding of them now, so iterating on my next contribution should be faster. I am still working out the issue with macos tests. I am on vacation next week, but I will pick it up again when I'm back.

File: test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_test.cc
Comment (alyssawilk) [inline test/extensions/filters/http/grpc_http1_reverse_bridge/reverse_bridge_test.cc:807]: can this be an integration test too?  for HTTP/2 I hope saying "content-length:20" sending 20 bytes and a reset would result in reset rather than "stream complete" but I'd like to have that tested :-)

**Review by alyssawilk — APPROVED** [review]
Thanks for all the extra tests here!

**Comment by snowp** [comment]
Can you merge main? I think that should resolve the CI failure

**Review by snowp — APPROVED** [review]
Thanks! Need a CI fix but LGTM

/wait

**Comment by baojr** [comment]
I'm still getting errors on what look like unrelated targets like "tcp_proxy_integration_test", so I am going to try merging main again.

**Comment by alyssawilk** [comment]
@htuch can you restamp?

**Review by htuch — COMMENTED** [review]
/lgtm api

**Pull request merged**
